multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,314 @@
+			r.labelQueue.Add(originalLabel)
+		} else {
+			// The original label still has other Pod that refers to it. Simply update the cache.
+			podNames.Delete(podNamespacedName)
+		}
+	}
+}
+
+// onPodCreateOrUpdate updates the Pod and label identity mapping in the cache, and
+// updates label identity ResourceExport if necessary (the Pod creation/update
+// event causes a new label identity to appear in the cluster or a label identity
+// no longer present in the cluster or both).
+func (r *LabelIdentityReconciler) onPodCreateOrUpdate(podNamespacedName, currentLabel string) {
+	// Based on the reconciler's predicate, the Pod has updated labels.
+	// Remove any original Pod-label mapping.
+	r.removeLabelForPod(podNamespacedName)
+
+	r.labelMutex.Lock()
+	defer r.labelMutex.Unlock()
+
+	r.podLabelCache[podNamespacedName] = currentLabel
+	podNames, ok := r.labelToPodsCache[currentLabel]
+	if !ok {
+		// Create a ResourceExport for this label as this is a new label.
+		klog.V(2).InfoS("Creating ResourceExport for label", "label", currentLabel)
+		r.labelToPodsCache[currentLabel] = sets.NewString(podNamespacedName)
+		r.labelQueue.Add(currentLabel)
+	} else {
+		// This is a seen label. Simply update the cache.
+		podNames.Insert(podNamespacedName)
+	}
+}
+
+// Run begins syncing of ResourceExports for label identities.
+func (r *LabelIdentityReconciler) Run(stopCh <-chan struct{}) {
+	defer r.labelQueue.ShutDown()
+
+	for i := 0; i < common.DefaultWorkerCount; i++ {
+		go wait.Until(r.labelQueueWorker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (r *LabelIdentityReconciler) labelQueueWorker() {
+	for r.processLabelForResourceExport() {
+	}
+}
+
+// Processes an item in the labelQueue. If syncLabelResourceExport returns an error,
+// this function handles it by re-queuing the item so that it can be processed again
+// later. If syncLabelResourceExport is successful, the label is removed from the queue
+// until we get notified of a new change. This function return false if and only if the
+// work queue was shutdown (no more items will be processed).
+func (r *LabelIdentityReconciler) processLabelForResourceExport() bool {
+	key, quit := r.labelQueue.Get()
+	if quit {
+		return false
+	}
+	defer r.labelQueue.Done(key)
+	err := r.syncLabelResourceExport(key.(string))
+	if err != nil {
+		// Put the item back on the workqueue to handle any transient errors.
+		r.labelQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Failed to sync ResourceExport for label identity", "label", key)
+		return true
+	}
+
+	// If no error occurs, we Forget this item so that it does not get queued again until
+	// another change happens.
+	r.labelQueue.Forget(key)
+	return true
+}
+
+// syncLabelResourceExport checks labelToPodsCache and determines whether a ResourceExport
+// needs to be created or deleted for the label identity.
+func (r *LabelIdentityReconciler) syncLabelResourceExport(label string) error {
+	r.labelMutex.RLock()
+	defer r.labelMutex.RUnlock()
```

Comment:
Locking the whole method will make the label sync workers pointless as all API calls will be made sequentially.
I think it only needs to lock when accessing labelToPodsCache.

Reply:
Thanks for catching this. Fixed